<a href="https://colab.research.google.com/github/seunghwan1228/Titanic/blob/master/5)_Titanic_ML_part_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# This document contains several ML models, 
the baseline of accuracy is 

Training : 0.8441011235955056  ≈  84.41%

Validation : 0.7988826815642458  ≈  79.89%
---
## Object : 

Running many different models and choose more valueable model to use 

---


Part 2
*  KNN
* Naive Bayes
* Logistic Regression
* SVM - linear / kernel


---

part 3
* ## Decision Tree - with visualization
* ## Boosting - Adaboost / gradient boost
* ## RandomForest
* ## Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM, AdaBoost, XGboost)



---


part 4
*  Neural network with Keras

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import VotingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import BernoulliNB, GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import roc_auc_score
from sklearn.metrics import roc_curve
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from xgboost import XGBClassifier
from sklearn.tree import export_graphviz
import graphviz


warnings.filterwarnings('ignore')

# 0) pre_steps

In [0]:
train = pd.read_csv('/content/drive/input/Titanic/working_dir/train_final.csv')

In [0]:
train.head()

0    1    2    3    4    5  ...   24   25   26   27   28  Survived
0 -0.502445  0.0  1.0  1.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0         0
1  0.786845  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0         1
2 -0.488854  0.0  1.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0         1
3  0.420730  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0         1
4 -0.486337  0.0  1.0  1.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0         0

[5 rows x 30 columns]

In [0]:
y = train.Survived
X = train.drop('Survived', axis=1)

In [0]:
X

In [0]:
print(X.shape)
print(y.shape)

(891, 29)
(891,)


In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state=42)

In [0]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(712, 29) (712,)
(179, 29) (179,)


# 1) Decision Tree

In [0]:
d_tree = DecisionTreeClassifier(random_state=42)

In [0]:
d_tree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=42, splitter='best')

In [0]:
d_tree.score(X_train, y_train)

0.949438202247191

In [0]:
d_tree.score(X_test, y_test)

0.7932960893854749

### Marker - overfitted

## Decision Tree - Visualizer

In [0]:
tree_visualizer = export_graphviz(d_tree, 
                                  class_names=['Non_Survived', 'Survived'], 
                                 rounded = True,
                                 filled = True)


graphviz.Source(tree_visualizer)

## Decision Tree Grid Search

In [0]:
# DecisionTreeClassifier() -- reference 

In [0]:
params = {'max_depth' : np.arange(2,20),
          'min_samples_split' : np.arange(2,50, 2),
          'max_leaf_nodes' : np.arange(2, 50, 2)
         }

In [0]:
decision_grid = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid=params, cv = 5)

In [0]:
decision_grid

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=DecisionTreeClassifier(class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort=False, random_state=42,
                                              splitter='best'),
             iid='warn', n_j...
             param_grid={'max_depth': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19]),
                  

In [0]:
decision_grid.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=DecisionTreeClassifier(class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort=False, random_state=42,
                                              splitter='best'),
             iid='warn', n_j...
             param_grid={'max_depth': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
       19]),
                  

In [0]:
decision_grid.best_params_

{'max_depth': 7, 'max_leaf_nodes': 12, 'min_samples_split': 14}

In [0]:
decision_grid.best_estimator_

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=7,
                       max_features=None, max_leaf_nodes=12,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=14,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=42, splitter='best')

In [0]:
decision_grid.best_score_

0.8286516853932584

In [0]:
d_tree_searched = decision_grid.best_estimator_

d_tree_searched.fit(X_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=7,
                       max_features=None, max_leaf_nodes=12,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=14,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=42, splitter='best')

In [0]:
d_tree_searched.get_params

<bound method BaseEstimator.get_params of DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=7,
                       max_features=None, max_leaf_nodes=12,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=14,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=42, splitter='best')>

In [0]:
d_tree_searched.score(X_train, y_train)

0.8539325842696629

In [0]:
d_tree_searched.score(X_test, y_test)

0.7932960893854749

## checking 5-Fold CV

In [0]:
cross_val_score(d_tree_searched, X_train, y_train, cv = 5)

array([0.81818182, 0.86013986, 0.83216783, 0.79577465, 0.83687943])

In [0]:
np.mean(cross_val_score(d_tree_searched, X_train, y_train, cv = 5))

0.8286287182001896

In [0]:
cross_val_score(d_tree_searched, X_test, y_test, cv = 5)

array([0.77777778, 0.69444444, 0.77777778, 0.66666667, 0.85714286])

In [0]:
np.mean(cross_val_score(d_tree_searched, X_test, y_test, cv = 5))

0.7547619047619047

## Conclusion - Decision Tree

for the Decision Tree, its performance is 

Train : 0.949438202247191  $\approx $ 94.94%

Test : 0.7932960893854749 $\approx $ 79.32%



---


with gridsearch : 

Train : 0.8539325842696629  $\approx $ 85.39%

Test : 0.7932960893854749 $\approx $ 79.32%


{'max_depth': 7, 'max_leaf_nodes': 12, 'min_samples_split': 14}



# 2) Ensemble - Boosting


*   Ada Boost
*   Gradient Boost
*   XG Boost


## 2-1 ) Ada Boost

In [0]:
ada_boost = AdaBoostClassifier(base_estimator= DecisionTreeClassifier(max_depth = 1),
                               random_state=42,
                               n_estimators=50,
                               learning_rate = 0.5)

In [0]:
ada_boost.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(class_weight=None,
                                                         criterion='gini',
                                                         max_depth=1,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort=False,
                                                         random_state=None,
                             

In [0]:
ada_boost.score(X_train, y_train)

0.8342696629213483

In [0]:
ada_boost.score(X_test, y_test)

0.7821229050279329

### 2-1-1 ) Ada Boost GridSearch

In [0]:
params = {'n_estimators' : np.arange(50, 1000, 50),
          'learning_rate' : [0.5, 1]}

In [0]:
ada_boost_mod = AdaBoostClassifier(DecisionTreeClassifier(max_depth = 1), random_state=42)

In [0]:
ada_grid = GridSearchCV(ada_boost_mod, param_grid=params, cv = 5)

In [0]:
ada_grid.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=AdaBoostClassifier(algorithm='SAMME.R',
                                          base_estimator=DecisionTreeClassifier(class_weight=None,
                                                                                criterion='gini',
                                                                                max_depth=1,
                                                                                max_features=None,
                                                                                max_leaf_nodes=None,
                                                                                min_impurity_decrease=0.0,
                                                                                min_impurity_split=None,
                                                                                min_samples_leaf=1,
                                                                                min

In [0]:
ada_grid.best_params_

{'learning_rate': 1, 'n_estimators': 300}

In [0]:
ada_grid.best_estimator_

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(class_weight=None,
                                                         criterion='gini',
                                                         max_depth=1,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort=False,
                                                         random_state=None,
                             

In [0]:
ada_best_mod = ada_grid.best_estimator_

In [0]:
ada_best_mod.fit(X_train, y_train)

AdaBoostClassifier(algorithm='SAMME.R',
                   base_estimator=DecisionTreeClassifier(class_weight=None,
                                                         criterion='gini',
                                                         max_depth=1,
                                                         max_features=None,
                                                         max_leaf_nodes=None,
                                                         min_impurity_decrease=0.0,
                                                         min_impurity_split=None,
                                                         min_samples_leaf=1,
                                                         min_samples_split=2,
                                                         min_weight_fraction_leaf=0.0,
                                                         presort=False,
                                                         random_state=None,
                             

In [0]:
ada_best_mod.score(X_train, y_train)

0.8679775280898876

In [0]:
ada_best_mod.score(X_test, y_test)

0.7932960893854749

## Conclusion - Ada_Boost

for the Ada Boost, its performance is 

Train : 0.83426966292134831  $\approx $ 83.43%

Test : 0.7821229050279329 $\approx $ 78.21%



---


with gridsearch : 

Train : 0.8679775280898876  $\approx $ 86.78%

Test : 0.7932960893854749 $\approx $ 79.32%


{'learning_rate': 1, 'n_estimators': 300}

## 2-2) Gradient Boost - sklearn

In [0]:
bg_clf = GradientBoostingClassifier(random_state=42)

In [0]:
bg_clf.fit(X_train, y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=100,
                           n_iter_no_change=None, presort='auto',
                           random_state=42, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [0]:
bg_clf.score(X_train, y_train)

0.9044943820224719

In [0]:
bg_clf.score(X_test, y_test)

0.7932960893854749

### 2-2-1) Gradient Boost Grid Search

In [0]:
params = {'n_estimators': np.arange(50, 1000, 50)}

In [0]:
bg_grid = GridSearchCV(GradientBoostingClassifier(random_state=42), param_grid=params, cv = 5)

In [0]:
bg_grid.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=GradientBoostingClassifier(criterion='friedman_mse',
                                                  init=None, learning_rate=0.1,
                                                  loss='deviance', max_depth=3,
                                                  max_features=None,
                                                  max_leaf_nodes=None,
                                                  min_impurity_decrease=0.0,
                                                  min_impurity_split=None,
                                                  min_samples_leaf=1,
                                                  min_samples_split=2,
                                                  min_weight_fraction_leaf=0.0,
                                                  n_estimators=100,
                                                  n_iter_no_change=None,
                                                  presort=

In [0]:
bg_grid.best_params_

{'n_estimators': 200}

In [0]:
bg_grid.best_score_

0.8314606741573034

In [0]:
bg_final_mod = bg_grid.best_estimator_

In [0]:
bg_final_mod.fit(X_train, y_train)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=200,
                           n_iter_no_change=None, presort='auto',
                           random_state=42, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [0]:
bg_final_mod.score(X_train, y_train)

0.922752808988764

In [0]:
bg_final_mod.score(X_test, y_test)

0.7988826815642458

## Conclusion - Gradient_Boost - sklearn

for the Gradient Boost, its performance is 

Train : 0.9044943820224719  $\approx $ 90.45%

Test : 0.7932960893854749 $\approx $ 79.32%



---


with gridsearch : 

Train : 0.922752808988764  $\approx $ 92.27%

Test : 0.7988826815642458 $\approx $ 79.89%

{'n_estimators': 200}

## 2-3) XG Boost (eXtreme Gradient Boosting)

In [0]:
xg_clf = XGBClassifier(random_state=42)

In [0]:
xg_clf.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [0]:
xg_clf.score(X_train, y_train)

0.8918539325842697

In [0]:
xg_clf.score(X_test, y_test)

0.8044692737430168

### 2-3-1) XGBoost Grid Search

In [0]:
params = {'gamma':[0, 0.001, 0.01, 0.1, 1, 10],
         'min_child_weight': [1, 3, 5, 7],
         'n_estimators': np.arange(50, 1001, 50),
         'max_depth' : [1,3,5,7,9]}

In [0]:
xg_grid = GridSearchCV(XGBClassifier(random_state=42), param_grid=params, cv = 5)

In [0]:
xg_grid.fit(X_train, y_train)

GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=XGBClassifier(base_score=0.5, booster='gbtree',
                                     colsample_bylevel=1, colsample_bynode=1,
                                     colsample_bytree=1, gamma=0,
                                     learning_rate=0.1, max_delta_step=0,
                                     max_depth=3, min_child_weight=1,
                                     missing=None, n_estimators=100, n_jobs=1,
                                     nthread=None, objective='binary:logistic',
                                     random_state=42, reg_alpha=0, reg_...
                                     scale_pos_weight=1, seed=None, silent=None,
                                     subsample=1, verbosity=1),
             iid='warn', n_jobs=None,
             param_grid={'gamma': [0, 0.001, 0.01, 0.1, 1, 10],
                         'max_depth': [1, 3, 5, 7, 9],
                         'min_child_weight': [1, 3, 5

In [0]:
xg_grid.best_params_

{'gamma': 0, 'max_depth': 3, 'min_child_weight': 7, 'n_estimators': 350}

In [0]:
xg_grid.best_score_

0.848314606741573

In [0]:
xg_final_mod = xg_grid.best_estimator_

In [0]:
xg_final_mod.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=7, missing=None, n_estimators=350, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [0]:
xg_final_mod.score(X_train, y_train)

0.8932584269662921

In [0]:
xg_final_mod.score(X_test, y_test)

0.7932960893854749

In [0]:
xg_grid.cv_results_

## Conclusion - XGboost 

for the XGBoost, its performance is 

Train : 0.8918539325842697  $\approx $ 89.19%

Test : 0.8044692737430168 $\approx $ 80.45%



---


with gridsearch : 

Train : 0.8932584269662921  $\approx $ 89.32%

Test : 0.7932960893854749 $\approx $ 79.32%

{'gamma': 0, 'max_depth': 3, 'min_child_weight': 7, 'n_estimators': 350}

# 3) Ensemble - RandomForest

In [0]:
from sklearn.ensemble import RandomForestClassifier

In [0]:
rfc_clf = RandomForestClassifier(random_state=42) # without tunning

In [0]:
rfc_clf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:
rfc_clf.score(X_train, y_train)

0.9410112359550562

In [0]:
rfc_clf.score(X_test, y_test)

0.8268156424581006

## RandomForest GridSearch

In [0]:
params = {'max_depth': np.arange(10, 100, 10), 
          'min_samples_split': [2,5,6,8,10], 
          'min_samples_leaf': [1,3,5,7],
          'n_estimators' : [10,50,100,150,200,250, 300, 350, 400, 450, 500, 600, 700, 800, 900, 1000]}

In [0]:
grid_search = GridSearchCV(RandomForestClassifier(random_state=42), 
                           param_grid=params, 
                           cv = 5, 
                           verbose = 1)

In [0]:
grid_search.fit(X_train, y_train)

Fitting 5 folds for each of 2880 candidates, totalling 14400 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 14400 out of 14400 | elapsed: 101.5min finished


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=RandomForestClassifier(bootstrap=True, class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators='warn', n_jobs=None,
                                              oob_score=Fa...dom_state=42,
                                              verbose=0, warm_start=False),
             iid='warn', n_jobs=None,
             param_grid={'max_d

In [0]:
grid_search.best_params_

{'max_depth': 10,
 'min_samples_leaf': 3,
 'min_samples_split': 10,
 'n_estimators': 10}

In [0]:
grid_search.best_score_

0.8328651685393258

In [0]:
g_rnd_clf = grid_search.best_estimator_

In [0]:
g_rnd_clf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=10, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=3, min_samples_split=10,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:
g_rnd_clf.score(X_train, y_train)

0.8553370786516854

In [0]:
g_rnd_clf.score(X_test, y_test)

0.8268156424581006

## Conclusion

for the RandomFrorest Classifier, 

train set : 0.925561797752809  $\approx $ 92.56%

test set : 0.8268156424581006 $\approx $ 82.68%


---


with gridsearch : 

Train :  0.8553370786516854  $\approx $ 85.53%

Test : 0.8268156424581006 $\approx $ 82.68%

{'max_depth': 10, 'min_samples_leaf': 3, 'min_samples_split': 10, 'n_estimators': 10}

# 4) Ensemble - Voting Classification



## 4-1) Hard voting for over validation accuracy 80% over on hard-voting
* Randomforest, KNN, Naive Bayes, Logit, SVM, XGboost, adaboost 

In [0]:
rnd_clf = RandomForestClassifier(random_state=42)
knn_clf = KNeighborsClassifier()
brl_clf = BernoulliNB()
log_clf = LogisticRegression(random_state=42)
svm_clf = SVC(random_state=42, probability=True)
ada_clf = AdaBoostClassifier(DecisionTreeClassifier(max_depth=1), random_state=42)
xgb_clf = XGBClassifier(random_state=42)

### 4-1H)
Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM, Adaboost, xgboost)

In [0]:
# Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM, Adaboost, xgboost)

vote_model = VotingClassifier(estimators= [('rnd', rnd_clf), ('knn', knn_clf), ('brl', brl_clf), ('log', log_clf), ('svm', svm_clf), ('ada', ada_clf), ('xgb', xgb_clf)])

In [0]:
vote_model.estimators

[('rnd',
  RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                         max_depth=None, max_features='auto', max_leaf_nodes=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, n_estimators='warn',
                         n_jobs=None, oob_score=False, random_state=42, verbose=0,
                         warm_start=False)),
 ('knn',
  KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                       metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                       weights='uniform')),
 ('brl',
  BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)),
 ('log',
  LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                     intercept_scaling=1, l1_ratio=None, max_iter=100,
                     multi_class='wa

In [0]:
vote_model.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model.score(X_train, y_train)

0.8581460674157303

In [0]:
vote_model.score(X_test, y_test)

0.8100558659217877

### 4-2H)
Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM)

In [0]:
# Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM)

vote_model = VotingClassifier(estimators= [('rnd', rnd_clf), ('knn', knn_clf), ('brl', brl_clf), ('log', log_clf), ('svm', svm_clf)])

In [0]:
vote_model.estimators

[('rnd',
  RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                         max_depth=None, max_features='auto', max_leaf_nodes=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, n_estimators='warn',
                         n_jobs=None, oob_score=False, random_state=42, verbose=0,
                         warm_start=False)),
 ('knn',
  KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                       metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                       weights='uniform')),
 ('brl',
  BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)),
 ('log',
  LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                     intercept_scaling=1, l1_ratio=None, max_iter=100,
                     multi_class='wa

In [0]:
vote_model.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model.score(X_train, y_train)

0.8469101123595506

In [0]:
vote_model.score(X_test, y_test)

0.7988826815642458

### 4-3H)
Ensemble - Vote classifier with (KNN, Naive Bayes, Logit, SVM)

In [0]:
vote_model2 = VotingClassifier(estimators= [('knn', knn_clf), ('brl', brl_clf), ('log', log_clf), ('svm', svm_clf)])


In [0]:
vote_model2

VotingClassifier(estimators=[('knn',
                              KNeighborsClassifier(algorithm='auto',
                                                   leaf_size=30,
                                                   metric='minkowski',
                                                   metric_params=None,
                                                   n_jobs=None, n_neighbors=5,
                                                   p=2, weights='uniform')),
                             ('brl',
                              BernoulliNB(alpha=1.0, binarize=0.0,
                                          class_prior=None, fit_prior=True)),
                             ('log',
                              LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1...
                                                 random_state=42, solver='warn',
   

In [0]:
vote_model2.fit(X_train, y_train)

VotingClassifier(estimators=[('knn',
                              KNeighborsClassifier(algorithm='auto',
                                                   leaf_size=30,
                                                   metric='minkowski',
                                                   metric_params=None,
                                                   n_jobs=None, n_neighbors=5,
                                                   p=2, weights='uniform')),
                             ('brl',
                              BernoulliNB(alpha=1.0, binarize=0.0,
                                          class_prior=None, fit_prior=True)),
                             ('log',
                              LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1...
                                                 random_state=42, solver='warn',
   

In [0]:
vote_model2.score(X_train, y_train)

0.8314606741573034

In [0]:
vote_model2.score(X_test, y_test)

0.7988826815642458

### 4-4H)
Ensemble - Vote classifier with (RandomForest, KNN, Naive Bayes, Logit)

In [0]:
# Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit)

vote_model3 = VotingClassifier(estimators= [('rnd', rnd_clf), ('knn', knn_clf), ('brl', brl_clf), ('log', log_clf)])

In [0]:
vote_model3.estimators

[('rnd',
  RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                         max_depth=None, max_features='auto', max_leaf_nodes=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, n_estimators='warn',
                         n_jobs=None, oob_score=False, random_state=42, verbose=0,
                         warm_start=False)),
 ('knn',
  KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                       metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                       weights='uniform')),
 ('brl',
  BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)),
 ('log',
  LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                     intercept_scaling=1, l1_ratio=None, max_iter=100,
                     multi_class='wa

In [0]:
vote_model3.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model3.score(X_train, y_train)

0.8735955056179775

In [0]:
vote_model3.score(X_test, y_test)

0.8100558659217877

### 4-5H)
Ensemble - Vote classifier with (RandomForest, Logit, svm)

In [0]:
# Ensemble - Vote classifier with (Randomforest, Logit)

vote_model4 = VotingClassifier(estimators= [('rnd', rnd_clf), ('log', log_clf), ('svm', svm_clf)])

In [0]:
vote_model4.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model4.score(X_train, y_train)

0.8426966292134831

In [0]:
vote_model4.score(X_test, y_test)

0.7988826815642458

### 4-6H)
Ensemble - Vote classifier with (RandomForest, xgboost)

In [0]:
vote_model5 = VotingClassifier([('rnd', rnd_clf), ('xgb', xgb_clf)])

In [0]:
vote_model5.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model5.score(X_train, y_train)

0.9087078651685393

In [0]:
vote_model5.score(X_test, y_test)

0.8100558659217877

###4-7H)
Ensemble - Vote classifier with (RandomForest, Logit, xgboost)

In [0]:
vote_model6 = VotingClassifier([('rnd', rnd_clf), ('xgb', xgb_clf), ('logit', log_clf)])

In [0]:
vote_model6.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model6.score(X_train,y_train)

0.8974719101123596

In [0]:
vote_model6.score(X_test, y_test)

0.8100558659217877

## 4-2)  Soft voting for over validation accuracy 80% over on hard-voting

In [0]:
rnd_clf = RandomForestClassifier(random_state=42)
knn_clf = KNeighborsClassifier()
brl_clf = BernoulliNB()
log_clf = LogisticRegression(random_state=42)
svm_clf = SVC(random_state=42, probability=True)
ada_clf = AdaBoostClassifier(DecisionTreeClassifier(max_depth=1), random_state=42)
xgb_clf = XGBClassifier(random_state=42)

In [0]:
vote_model = VotingClassifier(estimators= [('rnd', rnd_clf), ('knn', knn_clf), ('brl', brl_clf), ('log', log_clf), ('svm', svm_clf), ('ada', ada_clf), ('xgb', xgb_clf)], voting='soft')
vote_model3 = VotingClassifier(estimators= [('rnd', rnd_clf), ('knn', knn_clf), ('brl', brl_clf), ('log', log_clf)], voting='soft')
vote_model5 = VotingClassifier([('rnd', rnd_clf), ('xgb', xgb_clf)], voting='soft')

### 4.2S)
Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit, SVM, Adaboost, xgboost)

In [0]:
vote_model.fit(X_train,y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model.score(X_train, y_train)

0.8679775280898876

In [0]:
vote_model.score(X_test, y_test)

0.8100558659217877

### 4.2S)
Ensemble - Vote classifier with (Randomforest, KNN, Naive Bayes, Logit)

In [0]:
vote_model3.fit(X_train,y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model3.score(X_train, y_train)

0.8778089887640449

In [0]:
vote_model3.score(X_test, y_test)

0.8100558659217877

### 4.2S)
Ensemble - Vote classifier with (Randomforest,  xgboost)

In [0]:
vote_model5.fit(X_train, y_train)

VotingClassifier(estimators=[('rnd',
                              RandomForestClassifier(bootstrap=True,
                                                     class_weight=None,
                                                     criterion='gini',
                                                     max_depth=None,
                                                     max_features='auto',
                                                     max_leaf_nodes=None,
                                                     min_impurity_decrease=0.0,
                                                     min_impurity_split=None,
                                                     min_samples_leaf=1,
                                                     min_samples_split=2,
                                                     min_weight_fraction_leaf=0.0,
                                                     n_estimators='warn',
                                                     n_jobs=None,
      

In [0]:
vote_model5.score(X_train, y_train)

0.9269662921348315

In [0]:
vote_model5.score(X_test, y_test)

0.8212290502793296

## Conclusion - non-tunned model ensemble

for the hard-voting,  its highest performance is 
ensemble 

Train : 
between  0.8974719101123596 ~ 0.9087078651685393
$ \approx $ 89.19 ~ 90.87 %


Test : 0.8100558659217877 $\approx $ 81.00%



---

for the soft-voting (probability based)
vote ensembel (RandomForest, XGBoost) has heighest score as ,

Train : 0.9269662921348315  $\approx $ 92.70%

Test : 0.8212290502793296 $\approx $ 82.12%



## 4-3) Voting Classifier with tunned models

For this section, will use the tunned models of
* RandomForest
* XGBoost

Keras Document contains ensemble model with Keras, RandomForest, XGBoost

### 4-3-1) Hard Voting

In [0]:
xgb_mod = xg_grid.best_estimator_
xgb_mod

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=7, missing=None, n_estimators=350, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [0]:
rnd_mod = RandomForestClassifier(max_depth=10, min_samples_leaf=3, min_samples_split=10, n_estimators=10, random_state=42)
rnd_mod

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=10, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=3, min_samples_split=10,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)

In [0]:
tunned_vote = VotingClassifier(estimators=[('xgb', xgb_mod), ('rnd',rnd_mod)])

In [0]:
tunned_vote.fit(X_train, y_train)

VotingClassifier(estimators=[('xgb',
                              XGBClassifier(base_score=0.5, booster='gbtree',
                                            colsample_bylevel=1,
                                            colsample_bynode=1,
                                            colsample_bytree=1, gamma=0,
                                            learning_rate=0.1, max_delta_step=0,
                                            max_depth=3, min_child_weight=7,
                                            missing=None, n_estimators=350,
                                            n_jobs=1, nthread=None,
                                            objective='binary:logistic',
                                            random_state=42, reg_alpha=0,
                                            reg_lambda=1, scale_pos_weigh...
                                                     class_weight=None,
                                                     criterion='gini',
              

In [0]:
tunned_vote.score(X_train, y_train)

0.8721910112359551

In [0]:
tunned_vote.score(X_test, y_test)

0.8156424581005587

### 4-3-2) Soft Voting

In [0]:
tunned_vote_soft = VotingClassifier(estimators=[('xgb', xgb_mod), ('rnd',rnd_mod)], voting='soft')

In [0]:
tunned_vote_soft.fit(X_train, y_train)

VotingClassifier(estimators=[('xgb',
                              XGBClassifier(base_score=0.5, booster='gbtree',
                                            colsample_bylevel=1,
                                            colsample_bynode=1,
                                            colsample_bytree=1, gamma=0,
                                            learning_rate=0.1, max_delta_step=0,
                                            max_depth=3, min_child_weight=7,
                                            missing=None, n_estimators=350,
                                            n_jobs=1, nthread=None,
                                            objective='binary:logistic',
                                            random_state=42, reg_alpha=0,
                                            reg_lambda=1, scale_pos_weigh...
                                                     class_weight=None,
                                                     criterion='gini',
              

In [0]:
tunned_vote_soft.score(X_train, y_train)

0.8778089887640449

In [0]:
tunned_vote_soft.score(X_test, y_test)

0.7932960893854749

## Conclusion - tunned model ensemble

for the hard-voting,  its highest performance is 
ensemble 

Train : 0.8721910112359551 $\approx $ 87.22 %


Test : 0.8156424581005587 $\approx $ 81.56%



---

for the soft-voting (probability based)
vote ensembel (RandomForest, XGBoost) has heighest score as ,

Train : 0.8778089887640449  $\approx $ 87.78%

Test : 0.7932960893854749 $\approx $ 79.33%